src/MessageContainer.js

```diff
@@ -99,66 +82,68 @@ export default class MessageContainer extends React.Component {
   }
 
   scrollTo(options) {
-    this._invertibleScrollViewRef.scrollTo(options);
+    this.refs.flatListRef.scrollToOffset(options)
   }
 
-  renderRow(message, sectionId, rowId) {
-    if (!message._id && message._id !== 0) {
-      console.warn('GiftedChat: `_id` is missing for message', JSON.stringify(message));
+  renderRow({item,index}) {
+    if (!item._id && item._id !== 0) {
+      console.warn('GiftedChat: `_id` is missing for message', JSON.stringify(item));
     }
-    if (!message.user) {
-      if (!message.system) {
-        console.warn("GiftedChat: `user` is missing for message", JSON.stringify(message));
+    if (!item.user) {
+      if (!item.system) {
+        console.warn("GiftedChat: `user` is missing for message", JSON.stringify(item));
       }
-      message.user = {};
+      item.user = {};
     }
 
     const messageProps = {
       ...this.props,
-      key: message._id,
-      currentMessage: message,
-      previousMessage: message.previousMessage,
-      nextMessage: message.nextMessage,
-      position: message.user._id === this.props.user._id ? 'right' : 'left',
+      key: item._id,
+      currentMessage: item,
+      previousMessage: item.previousMessage,
+      nextMessage: item.nextMessage,
+      position: item.user._id === this.props.user._id ? 'right' : 'left',
     };
 
     if (this.props.renderMessage) {
       return this.props.renderMessage(messageProps);
     }
-    return <Message {...messageProps}/>;
-  }
-
-  renderScrollComponent(props) {
-    const invertibleScrollViewProps = this.props.invertibleScrollViewProps;
     return (
-      <InvertibleScrollView
-        {...props}
-        {...invertibleScrollViewProps}
-        ref={component => this._invertibleScrollViewRef = component}
-      />
-    );
+      <View style={{ transform: [{ scaleY: -1 },{perspective: 1280}]}}>
+        <Message {...messageProps}/>
+      </View>
+    )
   }
 
+  renderHeaderWrapper = () => {
+    return <View style={{ flex: 1, transform: [{ scaleY: -1 },{perspective: 1280}] }}>{this.renderLoadEarlier()}</View>;
+  };
+
+  _keyExtractor = (item, index) => item._id+" "+index
+
   render() {
     return (
       <View
         ref='container'
         style={styles.container}
       >
-        <ListView
+        <FlatList
           enableEmptySections={true}
           automaticallyAdjustContentInsets={false}
           initialListSize={20}
           pageSize={20}
-
+          ref='flatListRef'
+          keyExtractor={this._keyExtractor}
           {...this.props.listViewProps}
 
-          dataSource={this.state.dataSource}
+          data={this.state.messagesData}
 
-          renderRow={this.renderRow}
+          renderItem={this.renderRow}
           renderHeader={this.renderFooter}
-          renderFooter={this.renderLoadEarlier}
-          renderScrollComponent={this.renderScrollComponent}
+          renderFooter={this.renderLoadEarlier()}
+          style={{transform: [{scaleY: -1 },{perspective: 1280}]}}
```

Comment:
One last thing, what is the reasoning behind using this transform ourselves, and not using `inverted={true}`? I'm not sure why it not in the documentation (so you may very well know something I don't) :-)

Reply:
Inverted works only on RN 0.49 and above. Atleast for Huawei phones running Android 7

Reply:
Then maybe it is even a combination of the fact that inverted got added later + the perspective fix... Maybe good to wait some version before we use `inverted`, but still worth adding a TODO/NOTE?